### PR TITLE
Add FXIOS-8357 [v124] Login autofill improvements bottom sheet with login information

### PR DESCRIPTION
**Pull request by razvanlitianu**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8357)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/18497)

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift
Comment (razvanlitianu) [inline firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift:53]: I know we use mock usually for tests but I found them useful for SwiftUI previews for speed of development and compilation.

File: firefox-ios/nimbus-features/creditCardAutofillFeature.yaml
Comment (nbhasin2) [inline firefox-ios/nimbus-features/creditCardAutofillFeature.yaml:3]: @razvanlitianu any reason why we are removing CC feature? 🤔

Comment (nbhasin2) [reply firefox-ios/nimbus-features/creditCardAutofillFeature.yaml:3]: I see that you added [autofillFeature.yaml](https://github.com/mozilla-mobile/firefox-ios/pull/18866/files#diff-4ce029336fd9e9d18f5f74ae216bf35d302e2b205d4ab894d4b6848eeb6d283b) so I am guessing its taken care there

File: firefox-ios/Storage/Rust/RustLogins.swift
Comment (nbhasin2) [inline firefox-ios/Storage/Rust/RustLogins.swift:207]: Oh wow, so can nuke it fully?

File: firefox-ios/Client/Frontend/TabContentsScripts/LoginsHelper.swift
Comment (nbhasin2) [inline firefox-ios/Client/Frontend/TabContentsScripts/LoginsHelper.swift:148]: nit: would be nice to add it to the top of the class :)

File: firefox-ios/nimbus-features/creditCardAutofillFeature.yaml
Comment (razvanlitianu) [reply firefox-ios/nimbus-features/creditCardAutofillFeature.yaml:3]: Bad diffing, the file was renamed to accommodate all autofill experiments

File: firefox-ios/Client/Frontend/TabContentsScripts/LoginsHelper.swift
Comment (nbhasin2) [inline firefox-ios/Client/Frontend/TabContentsScripts/LoginsHelper.swift:21]: thank you for this!

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift
Comment (nbhasin2) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:2340]: nit: 
```suggestion
                // Keep showing bvc
                break 
```

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift
Comment (nbhasin2) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:2256]: @razvanlitianu since the feature flag has changed for credit card, did we update that too?

File: firefox-ios/nimbus-features/creditCardAutofillFeature.yaml
Comment (nbhasin2) [reply firefox-ios/nimbus-features/creditCardAutofillFeature.yaml:3]: ha, gotcha

**Comment by nbhasin2** [comment]
@lmarceau feel free to check out coordinator side of things :)

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (dataports) [inline firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:11]: I think you can remove the import

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (dataports) [inline firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:123]:
```suggestion
                guard let self else { return }
```

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (dataports) [inline firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:150]:
```suggestion
                guard let self else { return }
```

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (dataports) [inline firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:147]:
```suggestion
                router.dismiss(animated: true)
                parentCoordinator?.didFinish(from: self)
```

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (dataports) [inline firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:152]:
```suggestion
                parentCoordinator?.show(settings: .password)
                parentCoordinator?.didFinish(from: self)
```

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (dataports) [reply firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:123]: I've suggested some changes here and below to make the lines a bit more concise. These changes were introduced in Swift 5.8. 

You can read about it [here if you're interested](https://www.hackingwithswift.com/swift/5.8/implicit-self-weak-capture)

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (dataports) [inline firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:130]:
```suggestion
                    router.dismiss(animated: true)
                    parentCoordinator?.didFinish(from: self)
```

File: firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift
Comment (dataports) [inline firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift:6]: SwiftUI import is not needed here (@MainActor, ObservableObject and some of the other ones commonly used ones in SwiftUI VMs are available in Foundation) 

Personally, I try and avoid SwiftUI imports in view model classes just so view stuff doesn't accidentally get put in there.

**Comment by dataports** [comment]
SwiftUI looks good to me, just a couple small pieces of feedback. I'll let someone else hit approve though cause I don't feel like I'm the best person to review the rest of the code.

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (razvanlitianu) [reply firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:123]: Ah yes, old habits die hard 😅

File: firefox-ios/Storage/Rust/RustLogins.swift
Comment (razvanlitianu) [reply firefox-ios/Storage/Rust/RustLogins.swift:207]: Looks like it's not used anywhere else

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift
Comment (razvanlitianu) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:2264]: It didn't really change. Even tho the file name was changed the code is still compiled in one big swift file and the generated code is still the same.

File: firefox-ios/Client/Frontend/Autofill/LoginStorage.swift
Comment (lmarceau) [inline firefox-ios/Client/Frontend/Autofill/LoginStorage.swift:24]: I think this should be part of the `RustLogins.swift` file instead of being under the `Autofill` folder. This will easily be missed if someone else works on the RustLogins. The protocol you added can be directly in that file.

File: firefox-ios/Client/Nimbus/NimbusFeatureFlagLayer.swift
Comment (lmarceau) [inline firefox-ios/Client/Nimbus/NimbusFeatureFlagLayer.swift:188]: Indentation seems off 🤔

File: firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift
Comment (lmarceau) [reply firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift:53]: I think this should be in its own file rather than under the LoginListViewModel. If this is a pattern we will follow more (having mocks inside production code) I would have a specific folder for mocks maybe? In this PR there's `MockLogger` and `MockLoginStorage` that is introduced, as part of other files. Will be hard to discover and know they exists at their current location

Comment (lmarceau) [reply firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift:53]: Now thinking, will that have any conflicts with current existing `MockLogger` under the tests target? Should that one exists over there if we have one in the prod code? Feels like duplication.

File: firefox-ios/Client/Frontend/Autofill/LoginListView.swift
Comment (lmarceau) [inline firefox-ios/Client/Frontend/Autofill/LoginListView.swift:46]: nit; Feels like comments that are self implicit a bit?
```suggestion
```

File: firefox-ios/Client/Frontend/Autofill/LoginCellView.swift
Comment (lmarceau) [inline firefox-ios/Client/Frontend/Autofill/LoginCellView.swift:86]: Is that a button style that will be reused? If this is a button component style this potentially can go under the component library. CC @cyndichin

File: firefox-ios/Client/Frontend/Autofill/LoginCellView.swift
Comment (lmarceau) [inline firefox-ios/Client/Frontend/Autofill/LoginCellView.swift:73]: nit; same comment about those comments
```suggestion
```

File: firefox-ios/Client/Frontend/Autofill/LoginAutofillView.swift
Comment (lmarceau) [inline firefox-ios/Client/Frontend/Autofill/LoginAutofillView.swift:41]:
```suggestion
```

File: firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift
Comment (lmarceau) [inline firefox-ios/Client/Coordinators/CredentialAutofillCoordinator.swift:116]: Unit tests? 👀

File: firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift
Comment (lmarceau) [inline firefox-ios/Client/Coordinators/Browser/BrowserCoordinator.swift:566]: Unit tests? 👀

File: firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift
Comment (lmarceau) [inline firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift:37]: Unit tests? We even have the mocks implemented already to have tests for those functions 👀

File: firefox-ios/Client/Frontend/Autofill/LoginStorage.swift
Comment (razvanlitianu) [reply firefox-ios/Client/Frontend/Autofill/LoginStorage.swift:24]: LoginStorage is part of our target because we need a way to create mocks for SwiftUI previews. The conformation cannot happen there since RustLogins has no idea what LoginStorage is.

File: firefox-ios/Client/Frontend/Autofill/LoginCellView.swift
Comment (razvanlitianu) [reply firefox-ios/Client/Frontend/Autofill/LoginCellView.swift:86]: It can in theory because we're supposed to have the same cells for all the autofill screen but figma has, and I don't joke, 3 different versions of this cell so that's why for now it's a LoginButtonStyle. In the future I expect it to become AutofillStyle?

File: firefox-ios/Client/Frontend/Autofill/LoginStorage.swift
Comment (lmarceau) [reply firefox-ios/Client/Frontend/Autofill/LoginStorage.swift:24]: I'm not sure I understand. This protocol is just an async await enclosure on the Deferred method `public func listLogins() -> Deferred<Maybe<[EncryptedLogin]>>` no?

File: firefox-ios/Client/Frontend/Autofill/LoginCellView.swift
Comment (lmarceau) [reply firefox-ios/Client/Frontend/Autofill/LoginCellView.swift:86]: Designers should follow a standardized approach. We have components, Acorn and the design system being slowly put in place for those reasons. IMO the goal is that we have standardized cells and buttons, and not custom designs for one area of the application (in this case autofill). If that's not the case (this is what I understand from your comment?), you should discuss this with the designer as to what their plan is there.

File: firefox-ios/Client/Frontend/Autofill/LoginStorage.swift
Comment (razvanlitianu) [reply firefox-ios/Client/Frontend/Autofill/LoginStorage.swift:24]: `RustLogins.swift` is in the `Storage` framework. `LoginListViewModel` and `LoginAutofillView` are in our main target. I only need LoginStorage as a way to mock the logins for SwiftUI previews so I don't depend on Storage. If I move LoginStorage in the Storage framework then I will couple the views with a frameworks that does not need to know about any storage logic.

File: firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift
Comment (razvanlitianu) [reply firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift:53]: Same as the Storage, we need a way to mock the live implementation with mocks for SwiftUI previews. Having the mocks in test targets will not be accessible in the main target.

Comment (lmarceau) [reply firefox-ios/Client/Frontend/Autofill/LoginListViewModel.swift:53]: We discussed in DM and will leave this be for now

File: firefox-ios/Client/Frontend/Autofill/LoginCellView.swift
Comment (razvanlitianu) [reply firefox-ios/Client/Frontend/Autofill/LoginCellView.swift:82]: cc @jeffreygee @adudenamedruby

Comment (nbhasin2) [reply firefox-ios/Client/Frontend/Autofill/LoginCellView.swift:82]: @lmarceau Although I agree and from my last discussion we brought it up too @razvanlitianu if you remember. However this should not be a blocking change for this PR. 

Lets add a ticket for next part to address not just design as we also need to update JS component to use Autofill and not our old login scripts. So this work done is to support DMA in particular and reduce our sec risk (I can go into detail in DM) but there is a lot more work involved here in future.

Comment (lmarceau) [reply firefox-ios/Client/Frontend/Autofill/LoginCellView.swift:82]: Yeah no it's not a blocker. I chatted with Razvan in DM over this too. It's just a process thing we should go over/pick up as another task if needed as this doesn't seem to align with how we want UI components to be built long term.

**Review by lmarceau — APPROVED** [review]
Thank you! From my side the comments I put were resolved. There's just maybe some conversation to happen related to how the UI is built, to ensure we are aligned with the Componentization of our UI/aligned with the design system 🙏

File: firefox-ios/Client/Frontend/Autofill/LoginCellView.swift
Comment (nbhasin2) [reply firefox-ios/Client/Frontend/Autofill/LoginCellView.swift:82]: 100% agreed and thank you for bringing it up 🙏🏼 

@razvanlitianu I will merge this PR after BR run is good and lets use a future ticket to improve upon this

**Review by nbhasin2 — APPROVED** [review]

**Comment by nbhasin2** [comment]
image
no idea why BR failed

**Comment by nbhasin2** [comment]
@mergify backport release/v124

**Pull request merged**
